core/shared/src/main/scala/zio/prelude/fx/ZPure.scala

```diff
@@ -777,6 +777,53 @@ sealed trait ZPure[+W, -S1, +S2, -R, +E, +A] { self =>
 
   def tag: Int
 
+  /**
+   * Transforms ZPure to ZIO that either succeeds with `A` or fails with the first error `E`.
+   * State is read and written from/to a `ZRef`, log is written to another `ZRef`.
+   */
+  def toZIO[E1 >: E](s: ZRef[E1, E1, S2, S1], log: ZRef[E1, E1, Chunk[W], Any]): zio.ZIO[R, E1, A] =
```

Comment:
The ref for state is perfect. However, I would make this one return a tuple with the log in it. And maybe call it `toZIOWithLog` or `toZIOLogged`.

The other ones are perfect. 👍 ❤️

Reply:
But where would the log go? ZPure returns a log regardless if it fails or succeeds (`runAll: (Chunk[W], Either[Cause[E], (S2, A)])`).

Reply:
So can't we return the log in both the error and value channels since in either case there will be a log?

Reply:
What I wanted to say is that returning it via writing it to a Ref felt cleaner.
Alternative would be to return something like `ZIO[R, (Chunk[W], E1), (Chunk[W], A)]`.